### ISSUE TITLE
Integrate dataloader for ref/collection lazy loading

### ISSUE THREAD

**Issue by stephenh** [mikro-orm/mikro-orm#266]
I'm using mikro to implement a GraphQL server. In GraphQL, queries are not known ahead of time, and so the "endpoints"/resolvers are implemented very piecemeal, i.e. "get this project's clients", "okay now get the 1st client's name", "okay now get the 2nd client's name".

However, this can easily lead to N+1s where if a project has 100 clients, and the `Client` resolver's `name` function is called 100 times, on each call it does a `(await clientRef.load()).name`, we'll end up with 100 individual SQL calls.

The standard way of solving this in JS GraphQL implementations is to use dataloader (https://github.com/graphql/dataloader), which lets the programmer still implement the resolvers in the ^ fashion (I'm loading client 1, then client 2, then client 3) but behind the scenes it uses the JS event loop to turn each of the "load client 1", "load client 2", "load client 3" promises into a single meta-promise of "load clients 1, 2, 3" that then, once that promise is fulfilled with a batch-friendly "load all three" implementation (typically hand-written by the programmer), dataloader breaks up the list of "client 1, client 2, client3" back out into the individual promises, and gives it back to each call site, who know thinks "hey I just got a single client" but really in an N+1 safe way.

Granted, the scheme requires the GraphQL implementation to trigger the "load name for all 100 clients" in a single tick of the event loop, but this happens almost defacto by just calling the "load name" method from a regular for loop.

Anyway, b/c I need this "batch load refs" behavior, I've hand-written a dataloader that looks like:

```typescript
export class EntityDataLoader {
  private loader: DataLoader<Reference<any>, AnyEntity<any, any>>;

  constructor(em: EntityManager) {
    this.loader = new DataLoader<Reference<any>, AnyEntity<any, any>>(async refs => {
      // Group the refs per entity and mapValues to get their primary keys
      const idsPerEntity: Record<string, any[]> = {};
      asSequence(refs)
        .groupBy(ref => ref.__meta.className)
        .forEach((refs, key) => {
          idsPerEntity[key] = refs.map(ref => ref.__primaryKey);
        });
      // Load all the rows for each entity
      const promises = Object.entries(idsPerEntity).map(([entity, ids]) => {
        return em.getRepository(entity).find({ id: ids } as any);
      });
      await Promise.all(promises);
      // Now once the entities are already in our Unit Of Work, we can load the refs w/o triggering N SQL calls.
      return refs.map(async ref => await ref.load());
    });
  }

  load<T>(ref: Reference<T>): Promise<T> {
    // Our implementation promises to turn Ref<T> into a Promise<T> so this cast is okay.
    return this.loader.load(ref) as Promise<T>;
  }
}
```

So now in my `Client` resolver, it can look like:

```typescript
  Client: {
    async email(root) {
      return (await root.load()).email;
    },
    async name(root, args, ctx) {
      const client = await ctx.refDataLoader.load(root);
      return client.name;
    },
  },
```

And the `email` resolver will N+1 (because its directly calling `root.load()` where `root` is an `IdentifiedRef`) while `name` will not b/c it uses `ctx.refDataLoader.load` to batch load all of the refs accessed within this event loop.

It would be great if `IdentifiedRef.load()` (or more likely whatever internal loading logic it calls) used dataloader natively to just do this for me (i.e. by using a DataLoader instance associated with the Unit of Work).

This is also for refs, but this could generalize into any SQL-triggering operation, i.e. if I trigger a collection load on `client1.getAddresses()` and `client2.getAddresses()` (or what not, using the `Collection` API), instead of two SQLs for `from addresses where client_id = 1` and `...where client_id = 2`, a dataloader-aware `Collection` implementation would do this as `where client_id in (1, 2)` for me automatically.

In a way, this feature request is really a follow-up from my mini-rant on "preloads are not useful", insofar as 1) given I'm personally implementing a GQL server where it's really hard to know preloads, but I still need N+1 avoidance, I need dataloader-style batching loading, and 2) even though my case sounds "special" (oh, this is great for GQL servers), I assert its actually a generalized pattern, and could replace preloads all together, even for non-GQL endpoints/logic that "knows" (or _think_ it knows, until the code changes :-)) the right preloads to use.

I.e. something like:

```typescript
const project = loadProject(1);
for (const client in await project.clients.loadAll()) {
  for (const address in client.addresses.loadAll()) {
    /// do something with address
  }
}
```

Could be 3 SQLs of `select from project where id = 1`, `select from client where project_id in (...the project client ids...)`, and `select from address where client_id in (...the id of every client of the project...)`.

As a disclaimer, I know this is a big ask, and a large departure from the norm of "hint-driven preloads", but ironically the same "gotcha" of Node/JS ORMs suffering from "sorry, your object graph traversal of `project.getClients()` has to be a promise that you `await`" ends up providing the critical infrastructure ("we have to make a bunch of promises and then wait until the end of the event loop to do any I/O") needed for this sort of auto-batching/auto-flushing approach ("oh you want to do 10 separate I/O operations I'll combine those for you into 1 batch call").

Which, again ironically, I think in general is just pretty amazing b/c its framework-driven N+1 avoidance vs. programmer-driven N+1 avoidance, such that other languages/frameworks/ORMs have tried to provide the same thing, but without the Node/JS event loop trick, they end up generally being pretty complex and annoying i.e. https://github.com/facebook/Haxl or https://github.com/47deg/fetch.

**Comment by B4nan** [comment]
Wow, thanks for the detailed description. I must say I was recently thinking about something similar (as nextras orm, one of my inspirations, is actually doing the same afaik), although I did not know that it is called dataloader and it actually already exists :] I will need to spend more time to read this in more detail and go through the links, but generally I like the idea to support this in the ORM natively (maybe optionally for the beginning).

**Comment by emrosenf** [comment]
I have done a lot of work on this already. I have implemented dataloaders on all objections and relationships, and I have also ported [typeorm-loader](https://gitlab.com/Mando75/typeorm-graphql-loader) when I want to take advantage of the info parameter to do large joins instead of multiple queries.

Happy to compare

**Comment by lookfirst** [comment]
@emrosenf I'd love to see your port get into mikro somehow...

**Comment by vimmerru** [comment]
Hi! Could you explain how i can load level-2 or level-3 entities or avoid loading of big referenced collections with DataLoader?

Also i use a lot of pagination and filtering that requires joins. Can DataLoader help here?

**Comment by stephenh** [comment]
@vimmerru 

> load level-2 or level-3 references 

I'll selfishly be brief (ish), and link to something like: https://medium.com/@__xuorig__/the-graphql-dataloader-pattern-visualized-3064a00f319f (which is technically about GraphQL, but applies even to a non-GraphQL endpoint that is doing something like `for (friend : user.getFriends) { for (whatever : friend.getWhatever() { .. } }` (which I believe is the "level-2 or level-3" case you're talking about), a dataloader-driven approach will not load all 3 levels in 1 SQL call (which admittedly a preload hint could do), however it will at least use _only_ 3 SQL calls (1 to get the user, 1 to get the friends, and 1 to get the "whatever"s) by using larger and larger `where id in (...)` clauses.

So, it collapses what would be N sql calls (say 100 SQL calls to load "whatevers" for 100 friends) to 1 single SQL call "per level" (get all "whatevers" for all 100 friends in 1 SQL call).

Whether this level happens b/c of a GraphQL implementation detail, or nested levels of a REST endpoint's `for loop`, or being triggered dynamically within a lifecycle hook of 100 friends (as along as the ORM is calling `friendEntity.performWhateverLifeCycleHook()` within a single for `for (entity in entitiesToCallHooksOn` loop, i.e. a single event loop), then dataloader will "after the fact" combine each "level's calls" into a single call.

Which, given if implemented natively by the ORM, this it happens automatically & for-free (vs programmer-maintained preload hints), is a pretty great trade-off, i.e. 3 calls (or "1 call per-level") is not as perfect as 1 hint-/preload-driven SQL call, but it's still a lot better than N+1 SQL calls, so it's generally good enough.

(Or, pedantically, not even "good enough", it's "even better" b/c it also works in places that would otherwise generally don't/can't use preload-driven hints, i.e. lazy loading of refs/collections in lifecycle hooks, are now N+1 safe.)

> Also i use a lot of pagination and filtering that requires joins. Can DataLoader help here?

Well, dataloader can technically make application-level filtering more palatable (i.e. do a nested for loop and do filtering with JS code b/c you're nested for loops are N+1 safe), but otherwise no, pagination and filtering to _find the initial set of entities you're interested in_ is still a good idea, and then what dataloader helps with is auto-batching anything you do with those entities _after that_, instead of your initial "find across N tables with a join" _also_ having to have the right magical set of preload hints that line up with your post-initial-load business logic.

**Comment by darkbasic** [comment]
This is my take at this: https://github.com/mikro-orm/mikro-orm/pull/364
The main difference with @stephenh initial implementation is that this is properly typed and it also supports collections.

@emrosenf could you please share your dataloader implementation? I'm particulartly intrested in traversing the `info` object to make a single big query because that's what I was going to implement next.

**Comment by stephenh** [comment]
Fwiw @darkbasic aren't we essentially reimple Harusa or Postgraphile at that point?

I.e. info-based resolution into a single query (which they both do afaiu) generally requires your gql schema and DB schema to match 1-to-1. At which point, why bother with a domain layer like entity classes and mikro/typeorm/etc.?

Fwiw the approach I'm leaning towards/flushing out doesn't use info to "look ahead" into the next levels of the query, so it admittedly gives up on "a single query", but it does still lean on dataloader to "reactively" batch queries across each level, so if you touch three levels (parent/child/grandchildren) you'll have three queries instead of one, but that's still better than N.

And the upshot is that each level of the query can do it's own custom-if-needed logic on which entities are loaded/based on the field resolver impls.

(I haven't looked at your impl, but if you're shooting for info-based-one-query impl, you shouldn't need the dataloader library at all, b/c it's primarily for "auto-combining separate calls into 1 batch call", and the info approach is already essentially one DB call, afaiu.)

**Comment by darkbasic** [comment]
> I haven't looked at your impl, but if you're shooting for info-based-one-query impl, you shouldn't need the dataloader library at all, b/c it's primarily for "auto-combining separate calls into 1 batch call", and the info approach is already essentially one DB call, afaiu

My implementation is a pure general purpose dataloader, it isn't tied to GraphQL in any way and doesn't try to make a single big query, instead if tries to make a single query **per level**.

On the other hand I'm **also** experimenting on a GraphQL specific library which traverses the whole `info` object and makes a single, big query. But that's not something I think could/should be implemented into the core Mikro-ORM library, it's just an additional library which could be used when dealing with GraphQL projects.

**Comment by darkbasic** [comment]
I mentioned it just because @emrosenf said he already made something similar and so it could be a waste of time keep working on it.

**Comment by stephenh** [comment]
> doesn't try to make a single big query

Ah gotcha, sorry, I was scanning your PR on mobile and saw the term "single big query" and latched on to that. Didn't realize they were separate things. Makes sense!

**Comment by khayalan-mathew** [comment]
Native support for dataloader pattern would be really neat, was looking into mikroorm as a TypeORM alternative and already blown away by some of the feature, but this would be the cherry on top :)

**Comment by davidboom95** [comment]
This is my aproach for making a dataloader based on keys different than `id`
```
const entities = [...]
function generateComposedKeysLoaders(em: EntityManager){
    const composedKeyLoaders = {}
    const options = {
        cacheKeyFn: obj => JSON.stringify(obj)
    }
    entities.map(entity => {
        const entityRepo = em.getRepository(entity.name)
        composedKeyLoaders[`${entity.name}`] = new DataLoader(async (conditions) => {
            const entityRepo = em.getRepository(entity.name)
            const filters = {}
            conditions.map(condition => {
                Object.keys(condition).forEach(key => {
                    if(_.isArray(filters[key])) {
                        if(filters[key].includes(condition[key])) return
                        filters[key].push(condition[key])
                    } else {
                        filters[key] = [condition[key]]
                    }
                })
            })

            return await entityRepo.find(filters)
        }, options)
    })
    return composedKeyLoaders
}```

**Comment by jgiunta1** [comment]
This solution is more specific to type-graphql, but I wanted to post it here as well for anyone who needs it. 

https://github.com/MichalLytek/type-graphql/issues/51#issuecomment-911970003

**Comment by TheNightmareX** [comment]
It will be awesome to have this feature! It fits mikro's design well and benefits GraphQL apps greatly.

**Comment by bozorgmehr96** [comment]
Any update on this?

**Comment by darkbasic** [comment]
New PR for v6: https://github.com/mikro-orm/mikro-orm/pull/4321

**Comment by B4nan** [comment]
Closing as implemented in v6 via #4321
